resource_detectors/lib/opentelemetry/resource/detectors/container.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Resource
+    module Detectors
+      # Container contains detect class method for determining container resource attributes
+      module Container
+        extend self
+
+        UUID_PATTERN = '[0-9a-f]{8}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{4}[-_]?[0-9a-f]{12}'
+        CONTAINER_PATTERN = '[0-9a-f]{64}'
+        CONTAINER_REGEX = /(?<container>#{UUID_PATTERN}|#{CONTAINER_PATTERN})(?:.scope)?$/
+        CGROUP_V1_PATH = '/proc/self/cgroup'
+        CGROUP_V2_PATH = '/proc/self/mountinfo'
+
+        def detect
+          id = container_id
+          resource_attributes = {}
+
+          resource_attributes[OpenTelemetry::SemanticConventions::Resource::CONTAINER_ID] = id unless id.nil?
+          resource_attributes.delete_if { |_key, value| value.nil? || value.empty? }
+          OpenTelemetry::SDK::Resources::Resource.create(resource_attributes)
+        end
+
+        private
+
+        def container_id
+          [CGROUP_V2_PATH, CGROUP_V1_PATH].each do |cgroup|
+            next unless File.readable?(cgroup)
+
+            case cgroup
+            when CGROUP_V1_PATH
+              return container_id_v1
+            when CGROUP_V2_PATH
+              return container_id_v2
+            end
+          end
+
+          nil
+        end
+
+        def container_id_v1
+          File.readlines(CGROUP_V1_PATH, chomp: true).each do |line|
+            parts = line.split('/')
+            return parts.last if parts.last.match?(CONTAINER_REGEX)
+          end
+
+          nil
+        end
+
+        def container_id_v2
+          File.readlines(CGROUP_V2_PATH, chomp: true).each do |line|
+            parts = line.split('/')
+            parts.shift
+
+            parts.each do |part|
+              return part if part.match?(CONTAINER_REGEX)
```

Comment:
Thanks for your re-review @mwear.

My guess would be that filtering on `"hostname"` is done to limit the amount of regexmatching?
I initially included the filter on `"hostname"` as well; but decided to remove that; since it cannot be used reliably. e.g. when a user set the `hostname` themselves in [kubernetes](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#pod-v1-core) or [swarm](https://docs.docker.com/compose/compose-file/compose-file-v3/#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir); we would not be able to determine the `container.id` (although I think setting the `"hostname"` will probably be an edge case).

If we decide that we don't want to filter on `"hostname"` before regex matching; we might also remove `container_id_v1`; since `def container_id_v2` will also work for `cgroupv1` - although I still want to benchmark that before doing so - to keep the code a bit simpler. WDYT?

Reply:
I don't have a strong preference. I am most interested in the detector reliably and consistently detecting the container id. As far as performance goes, detecting the container id will be something that happens once at startup. While we don't want to give performance away unnecessarily, this is not a hot code path, so favoring simplicity is reasonable thing to do here. Let me know when you're happy with everything and I'll review it again.

Reply:
I've simplified the code a bit; according to me this should do the trick.